test/e2e/mtls_test.go

```diff
@@ -0,0 +1,180 @@
+	defer cleanup()
+
+	// Create envVars to be used in httpproxy app.
+	envVars := []corev1.EnvVar{{
+		Name:  targetHostEnv,
+		Value: name + "." + test.ServingNamespace + ".svc",
+	}}
+
+	t.Log("Creating a Knative Service for the httpproxy test app.")
+	names := test.ResourceNames{
+		Service: test.ObjectNameForTest(t),
+		Image:   "httpproxy",
+	}
+
+	test.CleanupOnInterrupt(func() { test.TearDown(clients, names) })
+	defer test.TearDown(clients, names)
+
+	resources, err := v1test.CreateServiceReady(t, clients, &names,
+		rtesting.WithEnv(envVars...),
+		rtesting.WithConfigAnnotations(map[string]string{
+			"sidecar.istio.io/inject": strconv.FormatBool(injectB),
+		}))
+	if err != nil {
+		t.Fatalf("Failed to create a Service: %v: %v", names.Service, err)
+	}
+
+	url := resources.Route.Status.URL.URL()
+	if _, err = pkgTest.WaitForEndpointState(
+		clients.KubeClient,
+		t.Logf,
+		url,
+		v1test.RetryingRouteInconsistency(pkgTest.MatchesAllOf(pkgTest.IsOneOfStatusCodes(expStatus))),
+		"HTTPProxy",
+		test.ServingFlags.ResolvableDomain,
+		test.AddRootCAtoTransport(t.Logf, clients, test.ServingFlags.Https),
+	); err != nil {
+		t.Fatalf("Get unexpected error: %v", err)
+	}
+
+}
+
+// In this test, we set up two apps: runtime and httpproxy.
+// runtime is a raw k8s pod and svc, while httpproxy is a knative service.
+// httpproxy is a proxy that redirects request to the runtime app.
+// The expected result is that the request sent to httpproxy app is successfully redirected
+// to runtime app when sidecar injection are both enabled or neither.
+func TestSvcToKubeSvc(t *testing.T) {
+	// testcases for table-driven testing.
+	var testCases = []struct {
+		name string
+		// injectA indicates whether istio sidecar injection is enabled for runtime (k8s) service.
+		// injectB indicates whether istio sidecar injection is enabled for httpproxy (knative) service.
+		injectA   bool
+		injectB   bool
+		expStatus int
+	}{
+		{"both-disabled", false, false, http.StatusOK},
+		{"a-disabled", false, true, http.StatusServiceUnavailable},
+		{"b-disabled", true, false, http.StatusBadGateway},
+		{"both-enabled", true, true, http.StatusOK},
+	}
+
+	t.Parallel()
+	cancel := logstream.Start(t)
+	defer cancel()
+
+	clients := Setup(t)
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			if os.Getenv("MESH") != "1" && (tc.injectA || tc.injectB) {
+				t.Skip("Skipping in no-mesh cluster")
+			}
+			cancel := logstream.Start(t)
+			defer cancel()
+			testSvcToKubeSvc(t, clients, tc.injectA, tc.injectB, tc.expStatus)
+		})
+	}
+}
+
+func createPodAndService(t *testing.T, clients *test.Clients, name string, inject bool) context.CancelFunc {
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: test.ServingNamespace,
+			Labels: map[string]string{
+				"test-pod": name,
+			},
+			Annotations: map[string]string{
+				"sidecar.istio.io/inject": strconv.FormatBool(inject),
```

Comment:
Not necessarily for this PR, but we should get rid of this. Should we move these tests into net-istio?

Reply:
Yeah, that is good idea, but someone may want to run other ingress with mesh env? For example, Kourier with Istio side-car. In that case, we still need to have this test here.
But I think we should discuss it on other place so I will open the issue for it.

Reply:
Actually that is the reason why I didn't put this test under `test/e2e/istio/`.

Reply:
I am not familiar with how other ingress enables mesh, but I think adding this annotation won't let all other ingress enable mesh, which may cause confusion.

My suggestion:
1. for this test, we get rid of istio annotations and put it under knative/serving
2. for the rest of test cases that need to configure this annotation, we put it under knative/net-istio. We could explore if there is a generic way to enable/disable mesh for different ingresses. If we find one, we can move those test cases back to knative/serving later.

WDYT?


WDYT?

Reply:
Yes, sounds good to me. Let me try to change it as you suggested.